Issue/12414 custom fields editor

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12414
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR is the first part in implementing the custom fields editor, it adds a new screen to edit existing custom fields, the same screen will be used when adding new fields too, and that should be taken into consideration when reviewing the code.

Many parts are still missing to allow keeping the PR size manageable, this includes the following:
1. Saving changes.
2. Returning results to the list screen and updating its state.
3. Using Aztec when editing HTML values.

**While the PR seems large, a big part of the changes is just tests and XML file changes due to using a nested nav graph, so it should be still easy to review.**

### Testing information
1. Open an order in wp-admin.
2. Tap on the "Custom Fields" section is visible (from the screen options at the top of the page)
3. Add some custom fields (for a more comprehensive testing, add fields that contain a URL, an email and a phone number)
5. Open the order in the app.
6. Tap on "View custom fields"
7. Tap on a custom field.
8. Make some changes.
9. Tap on the back button
10. Confirm a dialog is shown.
11. Test the behavior of the dialog buttons.

### The tests that have been performed
The above steps in addition to the validation rules.

### Images/gif
<img src="https://github.com/user-attachments/assets/a2842e36-e47d-4350-80d7-319182d5aa1f" width=360 />

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Create a new screen for the custom fields editor

hichamboushaba committed:
Update the custom field UI model to allow supplying custom key and value

hichamboushaba committed:
Add state logic to the ViewModel

hichamboushaba committed:
Introduce a nested navgraph for custom fields

hichamboushaba committed:
Add the editor to the nav graph

hichamboushaba committed:
Update the logic of the UI state

hichamboushaba committed:
Add initial UI of the editor screen

hichamboushaba committed:
Handle navigation from the list screen to the editor

hichamboushaba committed:
Make sure the secondary toolbar is hidden

hichamboushaba committed:
Fix logic of state initialization

hichamboushaba committed:
Handle the exit event

hichamboushaba committed:
Add logic to show a discard changes dialog

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>UiState</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Update validation rules to require a non-empty key and value

hichamboushaba committed:
Add tests

hichamboushaba committed:
Check if draft is not initialized yet to avoid overriding user changes

hichamboushaba committed:
Update logic to allow empty values

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/editor/CustomFieldsEditorViewModel.kt:
```diff
@@ -0,0 +1,136 @@
+        }
+    }
+
+    private fun calculateShowButtonState(draft: CustomFieldUiModel, stored: CustomField?): Boolean {
+        if (draft.key.isEmpty()) {
+            return false
+        }
+
+        return stored == null || stored.key != draft.key || stored.valueAsString != draft.value
+    }
```
The validation rules could change depending on the pending discussion p1724771478199229-slack-C03L1NF1EA3

But so far, the consensus is to disallow empty keys, but allow empty values, and that's what we do here, the `done` button will be hidden.

hichamboushaba committed:
Localize strings

hichamboushaba committed:
Make the editor screen scrollable

hichamboushaba committed:
Set the key field as single line

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/editor/CustomFieldsEditorViewModel.kt:
```diff
@@ -0,0 +1,136 @@
+        TODO()
+    }
+
+    private fun initState() {
```
Out of curiosity, is there a reason to have this single-use private function instead of adding this logic directly to the `init{}` block?

hichamboushaba replied:
Nothing major; I just wanted to make it more clear what the block does by the name `initState` instead of being a non-specific `init` logic.
Do you think having it inside the `init` block is easier to read?

hichamboushaba replied:
Oh, I just remembered, in an upcoming PR, this init logic will be totally removed (more about the change on the PR itself), so this block is going away.
Still, I would love to hear your thoughts on this.

JorgeMucientes replied:
> Still, I would love to hear your thoughts on this.

I understand the reasoning, and tbh I think is totally fine either way. It's really a minor np. 
But imho in this particular case, I don't think having the private method `initState` adds much to the code readability. I'd say 95% of the time, the code we add to the `init {}` block inside view models is state initialization anyway, and when there's more logic, it is when we start extracting functions out of it.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/editor/CustomFieldsEditorViewModel.kt:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    fun onBackClick() {
+        if (state.value?.showDoneButton == true) {
```
Np, there's one scenario where I believe we should show the discard changes dialog. When we edit the key value, we erase it. Even though it is not a state we can save because the key has to be non-empty, we have edited the actual value. 

Let me know what you think. This is a minor thing so feel free to ignore it.  


https://github.com/user-attachments/assets/5f4fe54f-d374-4b1e-8a43-aa20b929f10b

hichamboushaba replied:
Good point, this shows the logic here is broken, I'll try to improve it.

hichamboushaba replied:
I updated the logic with a new approach in a5d1fed368633cbf7806f15e3aba1c893442c82d, the state now has a property called `hasChanges` that checks for all changes regarldess or whether we will show the Done button or not, and then the `showDoneButton` is a computed property that checks if we have changes and also validates the state (basically it checks if `key` is not empty), WDYT about this @JorgeMucientes?

JorgeMucientes replied:
The new approach looks good to me. Thanks for addressing it 👍🏼

JorgeMucientes reviewed (approved):
Excellent job @hichamboushaba 👏🏼 
Everything works as expected. I just left one minor nitpick. Let me know what you think about it and feel free to merge without applying any change 👍🏼

hichamboushaba committed:
Update how changes are detected in the editor

This allows us to show the discard changes dialog even when we are not in a valid state for saving
